Add better error checking and testing

ajbozarth opened the pull request:
This PR add and updates the following:

- Better error handling
   - Web App doesn't start if model is running but returns errors on initial predictions
   - If the model goes down while the web app is running the user sees a error on upload but the web app still functions with the current content
- Add a pytest similar to the new MAX model pytest that does a curl call on the upload endpoint
- Updated Travis CI to start the model server and the web app server then run the pytest

This PR is paired with an update to the model Travis build (IBM/MAX-Image-Caption-Generator#14) that will trigger travis builds on this repo when there are new commits to the model repo. Due to this trigger I opted to do a local build of the model in travis rather than use docker hub since docker hub take ~10min to build images off of new commits

stevemart commented on requirements-test.txt:
```diff
@@ -0,0 +1,3 @@
+flake8
```
this is super opinionated and doesn't change anything functionally, but i believe `test-requirements.txt` is the usual name of the file, not `requirements-test.txt`. I can't find any docs on what pip handles but I've always used `test-requirements.txt` for my python projects.

a quick and dirty search on github: 
* https://github.com/search?q=test-requirements.txt&type=Code (47k hits)
vs
* https://github.com/search?q=requirements-test.txt&type=Code (5k hits)

ajbozarth replied:
I think I had seen this page (https://pypi.org/project/requirements-test.txt/) when naming that file and assumed it was the standard, I'll update it to match the more common naming.

stevemart replied:
https://pypi.org/project/requirements-test.txt/ was made (and last updated in 2016) by the pypa team to prevent squatters trying to install naughty python code on an easy typo.

stevemart commented on tests/test.py:
```diff
@@ -0,0 +1,30 @@
+import pytest
+import pycurl
+import io
```
can alphabetize the imports

ajbozarth replied:
nice catch, I figured flake. would've catch that

stevemart replied:
alphabetized imports are an extension: https://github.com/PyCQA/flake8-import-order

stevemart commented on tests/test.py:
```diff
@@ -0,0 +1,30 @@
+import pytest
+import pycurl
```
hmm? why not requests? http://docs.python-requests.org/en/master/

ajbozarth replied:
I copied the test from the model and edited it. It would probably be more consistent to use requirements though wouldn't it, I'll see what it would take to switch.

stevemart replied:
okay, i didn't realize it came from the model. whatever is easier for you.

MLnick replied:
Perhaps the model tests could look at using `requests` instead? cc @bdwyer2

ajbozarth replied:
I'm currently rewriting it with requests if you want to take my updates once I'm done and use them for the models

ajbozarth replied:
`requests` is the most "standard" http library for python as far as I've seen, it seems to me that `pycurl`s purpose is the replicate the curl command inside python rather than be a http library.

stevemart replied:
i'll admit i'm being biased :)

requests will handle a lot of the ugly stuff and in general make things cleaner/shorter: https://github.com/yudazilian/Pycurl-vs-Requests

requests also has a nice library for mocking requests: https://github.com/jamielennox/requests-mock should you ever want to mock things instead of testing against a live server.

stevemart commented on tests/test.py:
```diff
@@ -0,0 +1,30 @@
+import json
+
+
+def test_upload():
```
this does more than testing the upload, you're also testing an image and comparing it against an assumed value. suggest renaming this test to reflect what it's actually doing.

You could even split this up into something like the following and add more tests to test for errors (like a 400 for an invalid file format :) )

```
test_file = 'static/img/images/7-action-athlete-896567.jpg'
temp_file = 'static/img/images/MAX-7-action-athlete-896567.jpg'
baseball = 'a baseball player swinging a bat at a ball'

def baseTest():
    c = pycurl.Curl()
    b = io.BytesIO()
    c.setopt(pycurl.URL, 'http://localhost:8088/upload')
    c.setopt(pycurl.HTTPHEADER,
             ['Accept:application/json', 'Content-Type: multipart/form-data'])

def test_upload_baseball(baseTest):
    c.setopt(pycurl.HTTPPOST, [('file', (pycurl.FORM_FILE, test_file))])
    c.setopt(pycurl.WRITEFUNCTION, b.write)
    c.perform()
    assert c.getinfo(pycurl.RESPONSE_CODE) == 200
    c.close()
    response = b.getvalue()
    response = json.loads(response)[0]
    assert response['file_name'] == temp_file
    assert response['caption'] == baseball
```

ajbozarth replied:
I'll keep this in mind when updating to use requests

stevemart commented on app.py:
```diff
@@ -225,6 +240,9 @@ def main():
     end = time.time()
+    if error_raised:
+        logging.info("Failed to prepare metadata, stopping web server")
+        raise SystemExit
```
you could log the errors here for completeness, but it's already been raised

ajbozarth replied:
Yeah the error gets logged automatically by the requests functions so it would be redundant here. (the error actually gets thrown and logged for every thread, so it's really obvious)

stevemart replied:
ah okay. cool.

stevemart commented on .travis.yml:
```diff
@@ -2,5 +2,22 @@ language: python
+services:
+  - docker
+before_install:
+  - pushd ..
```
nice

ajbozarth commented:
I just pushed a major update to the tests. I switched to using `requests` instead of `pycurl` and added individual tests for each (non-html) endpoint's success and failure possibilities. There are parts of the tests file that may seem excessive since I prioritized making the test output clear while minimizing how often it git the web server for speed.

ajbozarth commented:
@bdwyer2 in the case of the model tests I can open a PR with the update to requests if you want me to, the one test you have is <5 line with requests.

ajbozarth commented:
@bdwyer2 I decided to just open a PR on the model with the switch to requests since it was super easy. 
https://github.com/IBM/MAX-Image-Caption-Generator/pull/16

ajbozarth commented on .travis.yml:
```diff
@@ -2,5 +2,22 @@ language: python
+  - docker build -t max-image-caption-generator .
+  - docker run -it -d -p 5000:5000 max-image-caption-generator
+  - popd
+  - sleep 60
```
I commented this on my other pr as well, the sleeps are to account for how long the model and web server's take to start. This length is definitely overkill, but I figured we'd rather have the tests never fail from starting too soon but take a littler longer to run. I can reduce this particular sleep to `30` though if you want

ajbozarth commented:
Based on Nick's comment, I changed `sleep` to `30` to match the model test, once the build passes I'll merge this since it has 2 approvals.
